Rewrite Listeners in terms of an endpoint concept.

jpeach opened the pull request:
This change explores one way to solve the problem described [here](https://docs.google.com/document/d/1UDxXWgDUZs5bHntons7bj2_5DUYpGfa_dMdxMX5jfRU),
which is that there is an implied endpoint abstraction that is awkwardly
split across the cluster operator and application owner roles.

Rewrite Listeners to model the concept of a logical network endpoint.
This change defines a Gateway more strongly as the network element to
which addresses can be attached, and Listeners as logical endpoints
within the network element.

The main goal of this change is to consolidate the name of an application
(i.e. the endpoint) into a single part of the API that is controlled
by one user persona. In this change, the application name is strictly
part of the Gateway spec and controlled by the cluster operator role,
while routes are controlled by the application owner.

The Listener protocol is now strongly defined to a core set of common
protocol stacks. This loss of flexibility improves clarity because it
makes it possible to specify how Listeners should be collapsed into
virtual hosts in an underlying proxy implementation.

This change doesn't update ListenerStatus, though clearly it no longer
makes complete sense. The most likely change needed for GatewayStatus
is to drop the Listeners field and fold listener errors directly into
the Conditions field.

jpeach commented:
/hold

robscott commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
+	// HTTPSProtocolType accepts HTTPS sessions.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TLSProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
```
I may be misunderstanding the scope of this PR, but on the surface these protocol additions feel out of scope.

jpeach replied:
Not fully. The problem is that certain behaviour can only be specified in terms of the protocol field. This means that, maybe, the open-ended protocol is a false abstraction and that in reality, the set of supported protocols is fixed.

robscott commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// Hostname must be supplied required for "HTTP" and "HTTPS" protocols, and may
+	// be supplied for "TLS" protocol.
```
The additional validation complexity here is an unfortunate consequence of this change.

jpeach replied:
I think that the complexity was already there, but this change lets us make it explicit. Previously, there was no *apparant* complexity exposed in the API because all the complexity of how to reconcile the hostname and TLS was implicit. The controller has to do *something* but the API provides no rules or other guidance.

robscott commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
 	//
 	// +optional
-	Address *ListenerAddress `json:"address,omitempty" protobuf:"bytes,2,opt,name=address"`
-	// Port is a list of ports associated with the Address.
+	// +kubebuilder:validation:MinLength=1
```
Are these two concepts compatible?

jpeach replied:
Maybe? What I'm trying to say is that the field is optional, but must not be the empty string.

robscott replied:
For strings, I think the common pattern would be to use a pointer here to indicate that it was optional. That way nil is valid and empty string is not. Since pointers can be a pain to work with, a better option might be to just document that an empty string will be ignored/treated as unspecified. I think that would mean removing the kubebuilder validation and instead relying on implementations to ignore empty string values here.

jpeach replied:
The hostname is now an optional struct, not a string.

robscott commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
-	// +optional
-	Protocol *string `json:"protocol,omitempty" protobuf:"bytes,4,opt,name=protocol"`
+	// +required
+	Port int32 `json:"port,omitempty" protobuf:"varint,3,opt,name=port"`
```
Under this model it seems like it would be helpful to have multiple ports/protocols per listener. I know we'd already discussed that as an option, but now that routes are per listener, it seems to be even more necessary.

jpeach replied:
There's no need for that, and I'd argue that it doesn't work. You can't have multiple processes accepting on the same socket unless you know the application protocol and that protocol has some method of discrimination. This means that you are, by definition, doing something protocol-specific, so trying to make that generic is a false abstraction that obscures rather than clarifies.

In this model, you can have a listener per port, and the rules for collapsing listeners onto the same underlying network port are clear. If you need more ports, then add more listeners with the same selector.

A further aspect that I didn't draw out in this PR is that the protocol and the route selector fields are tightly coupled and there have to be rules about how to handle this. Tying them together in the same struct makes it easier to elaborate the rules for handling that.

robscott replied:
Yep, as I've thought about this more it really does make sense to keep listeners unique per port/protocol combination.

jpeach commented:
> This an interesting concept, thanks for the work on it! The diff is relatively large, but if I'm understanding it correctly, the change entails moving hostname from routes to listeners and moving the route selector into listeners. Did I miss any other significant changes in the diff?

Nope, you have nailed the main intent. See attached doc for my reasoning that there is a "logical endpoint" object that we should model.

bowei commented:
/ok-to-test

bowei commented:
/assign

bowei commented:
Splitting the ports by listener seems to clean things up.
I'm not so sure about the hostname requirement:

How to have wildcard hostnames. 
This is a common use case (in fact we added support for it Ingress GA, maybe step backwards).

If not complex wildcards "*.foo.com", at least we at least need to be able to handle default hostname.

Given that with the listener + TLS + associated routes => you can valid the hostname matches and certs w/ all of the objects involved, can we remove hostname from the Gateway level?

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
+//
+// Listeners are considered compatible if the Protocol field is
+// "HTTP", HTTPS or "TLS", the Hostname field is specified, and the
+// Port field matches.
```
Is this defining compatibility (if yes, then compatibility with what) or validity of a Listener?

jpeach replied:
This is defining compatibility for the purposes of "collapsing compatible Listeners" in the preceding paragraph. The concept of collapsing Listeners is what allows virtual hosting).

jpeach replied:
I added some more text to specify the rules around collapsing Listeners into a virtual host configuration. I think it's clearer, but could still be better.

hbagdi commented:
> Given that with the listener + TLS + associated routes => you can validate the hostname matches and certs w/ all of the objects involved, can we remove hostname from the Gateway level?

That works for HTTP but not HTTPS.

I think this change is a positive one and we should proceed.
I have the same thought as @bowei on the hostname. Default and wildcard domains need to be supported here.
Here is what I think might work:

Keep `Hostname` in HTTPRoute. This is used for route matching.
In Gateway's Listener, we keep a field `AllowedHostnames []string` which acts as a filter.
Cluster Operator populate the field after they do the due diligence on their part (ensure DNS setup, TLS certs, approvals, etc).
App owners are now free to bind routes as they see fit as long as the hostname they wish to use is allowed in the Gateway Listener. If not, then they go and talk to the Cluster Operator.

`AllowedHostnames` can be strict, meaning each and every domain needs to be added, semi-strict by allowed a wildcard set of domains, or relaxed, where it allows for all domains (the easy path).

jpeach commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
+	// Listeners associated with this Gateway. Listeners define what addresses
+	// are bound on this Gateway. If no Listeners are provided, the GatewayClass
+	// may associate the Gateway with an address in an implementation-defined
+	// manner.
```
Yeh, this comment is a bit of a mess. I'll update to remove the address verbiage.

jpeach replied:
Removed.

jpeach commented:
> Splitting the ports by listener seems to clean things up.
> I'm not so sure about the hostname requirement:
> 
> How to have wildcard hostnames.
> This is a common use case (in fact we added support for it Ingress GA, maybe step backwards).

I need to go dig up and read the Ingress spec for this! I changed the Listener hostname to be a match type, so you can express wildcards as a domain match. The default would be an exact host name match.

> We at least need to be able to handle default hostname.

I documented the case where the Listener can omit the Hostname. This would allow your to specify routes without pinning to any name.

This, together with domain matching, required a fair amount of new text to specify how Listeners should be collapsed into a virtual host configuration.

> Given that with the listener + TLS + associated routes => you can validate the hostname matches and certs w/ all of the objects involved, can we remove hostname from the Gateway level?

Yeh I agree with @hbagdi on this. You need some deterministic information about the hostname in the Listener to do the right certificate selection. In the Google doc linked from the initial comment I discussed some of the corner cases here. Keeping some hostname spec in the Listener means that the semantics are consistent across changes in what the route selector selects.

jpeach commented:
> > Given that with the listener + TLS + associated routes => you can validate the hostname matches and certs w/ all of the objects involved, can we remove hostname from the Gateway level?
> 
> That works for HTTP but not HTTPS.
> 
> I think this change is a positive one and we should proceed.
> I have the same thought as @bowei on the hostname. Default and wildcard domains need to be supported here.
> Here is what I think might work:
> 
> Keep `Hostname` in HTTPRoute. This is used for route matching.
> In Gateway's Listener, we keep a field `AllowedHostnames []string` which acts as a filter.
> Cluster Operator populate the field after they do the due diligence on their part (ensure DNS setup, TLS certs, approvals, etc).
> App owners are now free to bind routes as they see fit as long as the hostname they wish to use is allowed in the Gateway Listener. If not, then they go and talk to the Cluster Operator.
> 
> `AllowedHostnames` can be strict, meaning each and every domain needs to be added, semi-strict by allowed a wildcard set of domains, or relaxed, where it allows for all domains (the easy path).

I restored the `Hostname` in HTTPRoute as per this suggestion. Probably we need some more explanatory text to describe how we intend it to work.

To address the default and wildcard cases, I borrowed the match concept from routes and change the Listener hostname to have "domain" and "exact" match types. A hostname can be omitted, which would give the Listener default route semantics.

costinm commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,184 @@ type GatewayList struct {
+	// Listeners associated with this Gateway. Listeners define logical
+	// endpoints that are bound on this Gateway. If no Listeners are
+	// provided, the GatewayClass may associate the Gateway with
+	// an address but the Gateway will not be able to accept any
```
Why not be able to accept ? If you don't specify any listener it could default to HTTP/HTTPS.

jpeach replied:
If there are no Listeners, there's no way to know where to send the traffic.

hbagdi replied:
If there are no listeners, you don't even know which port to listen on to accept any traffic in the first place.

jpeach replied:
I can resolve this by making Listeners required?

bowei replied:
Yes -- that's probably easiest rather than trying to figure out what empty listeners mean (for now)

jpeach commented:
Added some example YAML.

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
```
The default of `Exact` seems ambiguous. My understanding here is that if the value of the match is is empty, that implies a match against `*`, meaning all hosts are accepted.

Should the default then be:
```yaml
match: domain
value: "*"
```

jpeach replied:
The idea is that the `Hostname` field specifies a match and that the `Name` field in `HostnameMatch` must not be empty. The way to match all traffic (i.e. route to all hosts) is to omit the `Hostname` field altogether. I'll update the language to try and clarify this.

hbagdi replied:
Looks good now, resolving this thread.

hbagdi replied:
> I think it's clearer in an enumeration to have an explicit HostnameMatchAny and make that the default value, instead of empty

+1. This seems more readable and user-friendly.

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Value string `json:"value" protobuf:"bytes,2,name=value"`
```
Is `name` a better name here? `value` seems way too generic.

jpeach replied:
Sold!

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+// "HTTP", "HTTPS" or "TLS", the Hostname field is specified, and the Port
+// field matches. As a special case, each group of collapsed Listeners may
+// contain up to one Listener that omits the Hostname field. Listeners that
+// specify identical Hostname fields must not be treated as collapsible.
```
> contain up to one Listener that omits the Hostname field.

Why this limitation? 

> Hostname fields must not be treated as collapsible.

> specify identical Hostname fields must not be treated as collapsible.

Which one takes precedence?

jpeach replied:
I updated the text to be a bit more explanatory, but the basic idea is that you have a group of Listeners that you are going to collapse into a single acceptor configuration. Within this group, each Hostname field must be unique (i.e. two Listeners can't match the incoming hostname the same way), and you can only have one default Listener (i.e. only omit the Hostname from one Listener).

Both of these conditions are necessary to avoid ambiguous configurations where a controller would otherwise be required to merge Listeners in some way.

hbagdi replied:
https://github.com/kubernetes-sigs/service-apis/pull/193#issuecomment-635665805 explains the gap in my understanding which is why I added the above comment.

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+// If the GatewayClass collapses Listeners, the Hostname field must be matched
+// in order of most to least specific. "Exact" matches must be processed before
+// "Domain" matches, which must be processed before Listeners with no Hostname
+// field.
```
I don't understand any of this. Can you give an example?

The way I was thinking about collapsing is:
Multiple listeners on the same port, but with different match rules are collapsed
While evaluating the HTTPRoutes to satsify, the controller only satisfies a rout, if the hosts are valid based on the listener that the route is tied to.

Meaning, if you have listenerA with host `foo.com` and listenerB with `bar.com`, a route associated with listenerA that contains the host `bar.com` will be ignored. But if the same route is associated with listenerB, then it will be accepted.

jpeach replied:
When the client provides a hostname, the proxy has to match the incoming hostname using the Hostname matches from the collapsed group of Listeners. The order of matching should be from most to least specific.

jpeach commented:
> These are good changes, James! Thank you for working on this.
> 
> One downside is that the user needs to specify the `hostnames` twice for `http` and `https`.
> I think that this is a feature and not a limitation and we should move ahead and come back to this later (if users complain about it).

Yeh, there's a few configurations that you could argue have some textual redundancy, but on the bright side, they will be quite explicit :)

I've updated the test to try to clarify some of the questions inline, but left the conversations unresolved.

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
```
As a user, `Domain` seems to be as confusing as `Local` and `Cluster` external traffic policies, `domain` is so generic that it can mean different things to different people.
How about `Wildcard`?

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
 
-// Listener defines a
+// HostnameMatch specifies how a Listener should match the host name from a
+// client request. Depending on the incoming protocol, the match may apply to
```
Should we do `the match must apply` and then relax this later if need be?

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
-	// [1] https://tools.ietf.org/html/rfc1035
+	// Incoming requests are matched against Hostname before processing
+	// HTTPRoute rules. For example, if the HTTP request header
+	// contains "Host: foo.example.com", an Listener with Hostname value
```
typo: a Listener

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// As Listener that omits this field may be collapsed with other
```
typo: A Listener

hbagdi commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
 	// +optional
-	Address *ListenerAddress `json:"address,omitempty" protobuf:"bytes,2,opt,name=address"`
-	// Port is a list of ports associated with the Address.
+	Hostname *HostnameMatch `json:"hostname,omitempty" protobuf:"bytes,1,opt,name=hostname"`
```
Thinking more about this, why is this not an array instead?
Use case:
I want to control `*example.com` and `*.new.example.com` using a single Gateway.

hbagdi replied:
Reading through the examples I see how this would work. I'll post my comment in a comment on the PR itself rather than starting the conversation here.

hbagdi commented on examples/multiple-https.yaml:
```diff
@@ -0,0 +1,37 @@
+    value: auto-assign
+  listeners:
+  - hostname:
+      value: httpbin.example.com
```
This should be `name`. Same for other examples.

hbagdi commented on examples/wildcard-https.yaml:
```diff
@@ -0,0 +1,27 @@
+  listeners:
+  - hostname:
+      match: Domain
+      value: example.com
```
Should this be `*.example.com`?

jpeach replied:
It's not intended to be. For host name `foo.example.com` the domain is `example.com`. I assumed that requiring the leading `*.` didn't make things clearer :)

hbagdi commented:
James,
My understanding was flawed until this point.

What we are doing in this PR is:
```yaml
 listeners:
  - hostname:
      value: httpbin.example.com
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - resource: secrets
        name: httpbin
    routes:
    - routeSelector:
        matchLabels:
          app: httpbin
  - hostname:
      value: conformance.example.com
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - resource: secrets
        name: conformance
    routes:
    - routeSelector:
        matchLabels:
          app: conformance
```

And how I was (incorrectly) thinking about this was:

```yaml
 listeners:
  - hostnames:
      name: httpbin.example.com
      name: conformance.example.com
    port: 443
    protocol: HTTPS
    tls:
      - snis:
        httpbin.example.com
        certificateRefs:
        - resource: secrets
          name: httpbin
      - snis:
        conformance.example.com
        certificateRefs:
        - resource: secrets
          name: conformance
    routes:
    - routeSelector:
        matchLabels:
          app: example.com
```

The differnce in the tls config stands out the most but that's less important.

The important part is about having a `routeSelector` that can be shared across hostnames. Why?
1. If you want to share the route selector for multiple domains, that is possible in this model. In the model where route selector is per hostname/port pair, if you use a common selector, then a HTTPRoute resource will be considered invalid for one listener (where hostname doesn't match) but considered valid for another listener (where hostname does match), This can get confusing and a little cumbersome to use.
2. In case a user wants different route selector, that's still possible, you can either create another gateway resource and specify a different route selector for that, or create another listener in the same gateway to achieve that.

jpeach commented:
> The important part is about having a `routeSelector` that can be shared across hostnames. Why?
> 
> 1. If you want to share the route selector for multiple domains, that is possible in this model. In the model where route selector is per hostname/port pair, if you use a common selector, then a HTTPRoute resource will be considered invalid for one listener (where hostname doesn't match) but considered valid for another listener (where hostname does match), This can get confusing and a little cumbersome to use.

I don't think this is quire right. The spec implicitly describes 2 layers of matching for hostnames. The first applies at the Listener, the second at the HttpRoute. The way this PR specs it, you wold have 2 Listeners for "a.com" and "b.com", and give both Listeners the same route selector. That doesn't mean that routes would ever be invalid for a listener, just that the hostname for some routes would never be matched by requests that come from some Listeners.

Basically, there's no requirement that the Hostname in the `HTTPRouteHost` has to match anything in any Listener. It will just either work or not.

> 2. In case a user wants different route selector, that's still possible, you can either create another gateway resource and specify a different route selector for that, or create another listener in the same gateway to achieve that.

We already have multiple Listeners and rules around collapsing so I'm a bit reluctant to make the matching more complex too. If Listeners have multiple hostname matches, we have to order them and figure out how they might conflict.

We should be flexible, but also avoid having too many ways to say the same thing.

hbagdi commented:
I think what you are saying makes sense now. Thanks for being patient with me and explaining this.
I can buy everything you have said.

I've some new questions:
0. If an HTTPRoute `c.com` is created, to which traffic is never routed, the status of the route needs to be Invalid to signal to the user that something is amiss rather than "this route will never be matched".
1. How/Where to define a default TLS cert-key pair?
2. In the case where the hostname match is not defined for a listener (match all domains), how are TLS configs defined?

jpeach commented:
> I think what you are saying makes sense now. Thanks for being patient with me and explaining this.
> I can buy everything you have said.
> 
> I've some new questions:
> 0. If an HTTPRoute `c.com` is created, to which traffic is never routed, the status of the route needs to be Invalid to signal to the user that something is amiss rather than "this route will never be matched".

I'm not sure whether I'd classify this is an error, it probably needs it's own issue to discuss. Controllers could add a warning condition on the Gateway as a quality of implementation improvement, but maybe the Listener becomes invalid if the Controller can prove it can't actually route any traffic?

> 1. How/Where to define a default TLS cert-key pair?
> 2. In the case where the hostname match is not defined for a listener (match all domains), how are TLS configs defined?

If a Listener omits the hostname field, it can still have port, protocol, and tls fields. The spec describes how each collapse group can have one of these types of Listener, which captures all unmatched traffic. This is also how you would define a default TLS certificate.
> Edit:
> Re-reading the PR, it seems that you haven't touched the TLS parts and we should probably keep the TLS conversation out of scope from this PR. We can revisit how to structure the `TLSConfig` part in a subsequent PR.

Yup, that was intentional :)

bowei commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+	// that define this concept.
 	//
-	// [1] https://tools.ietf.org/html/rfc1035
+	// Incoming requests are matched against Hostname before processing
```
I think we can put how it matches w/ the given types aboe.

Requests will be matched according to the given HostnameMatch type defined above...

jpeach replied:
Rewritten a bit.

bowei commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
-	// Support: Core
+	// A Hostname match should be supplied for "HTTP" and "HTTPS"
+	// protocols, and may be supplied for "TLS" protocol. If no
+	// match is supplied, then this Listener must accept all client
```
Can use more generic language here:

if hostname is specified, it will be matched according to the semantics of the protocol, describe behavior in the protocol enum to avoid duplication.

Hostname is optional? (you have should and may here)

to be accurate, the connection is accepted  but will result in 404ish result

jpeach replied:
> to be accurate, the connection is accepted but will result in 404ish result

Can you clarify what text you are referring to here?

bowei replied:
To get the request at all, the connection will be accept()'ed, even if the TLS handshake is bad, host is wrong etc.
I'm referring to 

"if no match is supplied, then this Listener must accept all client *connections*"

jpeach replied:
Oh I see. The new language is "accepts all client traffic", but this doesn't mean it proxies it, just that the listener gets the packets. It might do anything with them.

bowei commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// A Listener that omits this field may be collapsed with other
```
I think we should be more explicit and use the HostMatchAny?

hbagdi commented:
> I'm not sure whether I'd classify this is an error, it probably needs it's own issue to discuss. Controllers could add a warning condition on the Gateway as a quality of implementation improvement, but maybe the Listener becomes invalid if the Controller can prove it can't actually route any traffic?

Let's discuss this in another issues or a subsequent PR. I'd like to see this PR being merged and iterated upon rather than having more blockers on it.

> If a Listener omits the hostname field, it can still have port, protocol, and tls fields. The spec describes how each collapse group can have one of these types of Listener, which captures all unmatched traffic. This is also how you would define a default TLS certificate.

But in that case, there is no way to define different cert-key pairs for different domains. That's a limitation of TLSConfig and not introduced in this PR. I've something in mind and I'll PR once this is merged in.

jpeach commented:
> > If a Listener omits the hostname field, it can still have port, protocol, and tls fields. The spec describes how each collapse group can have one of these types of Listener, which captures all unmatched traffic. This is also how you would define a default TLS certificate.
> 
> But in that case, there is no way to define different cert-key pairs for different domains. That's a limitation of TLSConfig and not introduced in this PR. I've something in mind and I'll PR once this is merged in.

The way do this is to have a Listener for each domain that you have TLS material for. The Hostname file in the Listener can be matched to a domain and that Listener specified the TLS configuration.

jpeach commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+
+	// Name contains the name to match against. This value must be a
+	// fully qualified host or domain name as defined by
+	// [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
```
I pulled this language from elsewhere, but I've added some (hopefully) clarification that it is a registered name that is relevant here.

jpeach commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse Listener definitions into single
```
> Is this "May" or "Must"?

Currently MAY, because I didn't want to require implementations to do this. Maybe some gateway classes just don't implement layer 7 matching but should otherwise be conformant.

> Hard to see how a Gateway with a PortConflict can be implemented?

Not sure whether I'm interpreting your question correctly, but if you add 2 Listeners with matching Port fields and TCP protocol, that would be a PortConflict (since they are not collapsible).

> Can we use the term "compatible Listener" instead of collapsible?

Previously I used "compatible", but it was less specific than using "collapsible" as a term of art. Not sure what the right approach is here; using a more generic word makes it harder to mean something extremely specific.

> Should we just say that within a given Gateway, they must all be compatible?

Yes, but what language should we use to describe what compatibility means?

I'll try another pass at this in terms of compatibility and see whether that works out ...

jpeach commented:
> +1 to Wildcard.
>
> Exact is also a "domain"

I don't know why gitHub doesn't show me a field to reply inline, but 🤷 

I think that in the YAML, it reads fairly clearly:

Exact match:
```
  - hostname:
      match: Exact
      name: ssh.example.com
```

Domain match:
```
  - hostname:
      match: Domain
      name: example.com
```

bowei commented on api/v1alpha1/gateway_types.go:
```diff
@@ -23,7 +23,10 @@ import (
 
-// Gateway represents an instantiation of a service-traffic handling infrastructure.
+// Gateway represents an instantiation of a service-traffic handling
+// infrastructure. The distinguishing characteristics of a gateway are
```
Probably redundant? (network endpoint => IP address for the most part)

jpeach commented:
Updated with a number of changed to address review feedback, including the "Any" hostname match type.

I think I've captured most of the feedback. Two suggestions that I haven't addressed are
- s/domain/wildcard/
- s/collapsable/compatible/

bowei commented:
thanks -- are you going to incorporate those changes? I will merge it after that

jpeach commented:
> thanks -- are you going to incorporate those changes? I will merge it after that

I'm not too keen on the wildcard nomenclature. I'll experiment with rewriting the Listener collapse docs though.

jpeach commented:
> > thanks -- are you going to incorporate those changes? I will merge it after that
> 
> I'm not too keen on the wildcard nomenclature. I'll experiment with rewriting the Listener collapse docs though.

Updated the listener collapse docs to emphasize "compatible" listeners rather than "collapsible" listeners. I moved all that text up to the Gateway's listeners field since it is applicable to how a gateway should deal with the listeners array there.

jpeach commented:
Updated field tags and examples so that all examples can be applies to a cluster. Filed the following issues to track follow on work:

#206 Discuss domain match vs. wildcard match for listener hostnames kind/feature
#205 WebSocket protocol support kind/feature
#204 h2c (cleartext HTTP/2) protocol support kind/feature
#203 Protocol type extensibility kind/bug
#202 Re-address Gateway and Listener error conditions kind/bug

hanlins commented:
Can you change [gateway conditions](https://github.com/kubernetes-sigs/service-apis/blob/a0723b3e82a8fbe083f9914fd2b70d3b84764369/api/v1alpha1/gateway_types.go#L366-L371) since `route` is not part of `Gateway`? Also it would be nice to add `GatewayConditionType` for `Addresses` like `InvalidAddress`.

jpeach commented:
> Can you change [gateway conditions](https://github.com/kubernetes-sigs/service-apis/blob/a0723b3e82a8fbe083f9914fd2b70d3b84764369/api/v1alpha1/gateway_types.go#L366-L371) since `route` is not part of `Gateway`? Also it would be nice to add `GatewayConditionType` for `Addresses` like `InvalidAddress`.

There's a lot of conditions that don't make sense after this change, and also conditions that are missing. Since this PR is quite intrusive already, I'm proposing to spin that out into #202.

bowei commented on api/v1alpha1/gateway_types.go:
```diff
@@ -54,78 +55,212 @@ type GatewayList struct {
+	// definitions into single implementation-defined acceptor
+	// configuration even if their Port fields would otherwise conflict.
+	//
+	// Listeners are compatible if all of the following conditions are true:
```
We can fix this in a later commit, but it would be good to frame this from the user's standpoint, as the current description is somewhat hard to parse.

jpeach commented:
> On Jun 6, 2020, at 8:46 AM, Bowei Du <notifications@github.com> wrote:
> 
> ﻿
> @bowei commented on this pull request.
> 
> In api/v1alpha1/gateway_types.go:
> 
> > @@ -54,78 +55,212 @@ type GatewayList struct {
>  type GatewaySpec struct {
>  	// Class used for this Gateway. This is the name of a GatewayClass resource.
>  	Class string `json:"class" protobuf:"bytes,1,opt,name=class"`
> -	// Listeners associated with this Gateway. Listeners define what addresses,
> -	// ports, protocols are bound on this Gateway.
> +
> +	// Listeners associated with this Gateway. Listeners define
> +	// logical endpoints that are bound on this Gateway's addresses.
> +	// At least one Listener MUST be specified.
> +	//
> +	// Each Listener in this array must have a unique Port field,
> +	// however a GatewayClass may collapse compatible Listener
> +	// definitions into single implementation-defined acceptor
> +	// configuration even if their Port fields would otherwise conflict.
> +	//
> +	// Listeners are compatible if all of the following conditions are true:
> We can fix this in a later commit, but it would be good to frame this from the user's standpoint, as the current description is somewhat hard to parse.
> 

I’ve been through a number of revisions of this, but I still agree that it could be improved. Happy to take editorial suggestions or discuss on a call :)

bowei commented:
/approve

let's merge and iterate

robscott commented:
Just confirmed with @jpeach that we can remove the hold and get this in. 

/hold cancel
/lgtm
